### PR TITLE
feat:add a fields in opportunity item

### PR DESCRIPTION
**Pull request by nmimsnikhil**
-add a 'rate' and 'amount' fields in opportunity_item.
-total amount of items is displaying in opportunity_amount field.
task: https://bloomstack.com/desk#Form/Task/TASK-2020-01431
opportunity_item

### CODE REVIEW

File: erpnext/crm/doctype/opportunity/opportunity.js [hunk 1]
```diff
@@ -198,4 +198,28 @@ cur_frm.cscript.item_code = function(doc, cdt, cdn) {
...
+var calculate_amount = function(doc){
+	var items = doc.items || [];
+	var total_amount = 0;
+	for (var i=0; i < items.length; i++){
+		var amount = items[i].qty * items[i].rate;
+		frappe.model.set_value('Opportunity Item', items[i].name, 'amount', amount);
+		total_amount += amount;
+	}
+	cur_frm.set_value("opportunity_amount", total_amount);
+};
```
Comment (Alchez) [inline erpnext/crm/doctype/opportunity/opportunity.js]: Add this in the controller at the start of this file

Comment (Alchez) [inline erpnext/crm/doctype/opportunity/opportunity.js]:
```suggestion
calculate_amount: function(frm, cdt, cdn) {
	let total_amount = 0;
	frm.doc.items.forEach(item => {
		let amount = item.qty * item.rate;
		frappe.model.set_value(cdt, cdn, 'amount', amount);
		total_amount += amount;
	})
	frm.set_value("opportunity_amount", total_amount);
};
```

**Review by Alchez — CHANGES_REQUESTED** [review]

File: erpnext/crm/doctype/opportunity/opportunity.js [hunk 2]
```diff
@@ -6,6 +6,20 @@ frappe.provide("erpnext.crm");
 
 cur_frm.email_field = "contact_email";
 frappe.ui.form.on("Opportunity", {
+
+	calculate_amount: function (frm) {
+		let total_amount = 0;
+		console.log(frm);
```
Comment (Alchez) [inline erpnext/crm/doctype/opportunity/opportunity.js]: Remove the console logs

File: erpnext/crm/doctype/opportunity/opportunity.js [hunk 3]
```diff
@@ -6,6 +6,20 @@ frappe.provide("erpnext.crm");
...
+		frm.doc.items.forEach(item => {
+			let amount = item.qty * item.rate;
+			console.log(amount);
+			frappe.model.set_value(item.doctype, item.name, 'amount', amount);
+			total_amount += amount;
+			amount = 0;
```
Comment (Alchez) [inline erpnext/crm/doctype/opportunity/opportunity.js]: Is this not re-initializing this variable on every loop iteration?

File: erpnext/crm/doctype/opportunity/opportunity.js [hunk 4]
```diff
@@ -6,6 +6,20 @@ frappe.provide("erpnext.crm");
...
+		})
+		cur_frm.set_value("opportunity_amount", total_amount);
```
Comment (Alchez) [inline erpnext/crm/doctype/opportunity/opportunity.js]:
```suggestion
		frm.set_value("opportunity_amount", total_amount);
```

File: erpnext/crm/doctype/opportunity/opportunity.js [hunk 5]
```diff
@@ -198,4 +212,18 @@ cur_frm.cscript.item_code = function(doc, cdt, cdn) {
 			}
 		})
 	}
-}
\ No newline at end of file
+}
+
+frappe.ui.form.on("Opportunity Item", {
+	qty: function (frm) {
+		frm.trigger("calculate_amount")
```
Comment (Alchez) [inline erpnext/crm/doctype/opportunity/opportunity.js]: Add a semicolon at the end of all 3 calls

**Review by Alchez — CHANGES_REQUESTED** [review]

File: erpnext/crm/doctype/opportunity/opportunity.js (hunk 3 above)
Comment (nmimsnikhil) [reply erpnext/crm/doctype/opportunity/opportunity.js]: sorry that 'amount = 0' by mistake is in loop.

File: erpnext/crm/doctype/opportunity_item/opportunity_item.json [hunk 6]
```diff
@@ -1,469 +1,151 @@
...
+   "fetch_from": "item_code.valuation_rate",
+   "fieldname": "rate",
+   "fieldtype": "Currency",
+   "in_list_view": 1,
+   "label": "Rate",
+   "options": "currency"
+  },
+  {
+   "default": "0",
+   "fieldname": "amount",
```
Comment (Alchez) [inline erpnext/crm/doctype/opportunity_item/opportunity_item.json:132]: Make this read-only, it shouldn't be editable

File: erpnext/crm/doctype/opportunity_item/opportunity_item.json [hunk 7]
```diff
@@ -1,469 +1,151 @@
...
+   "hidden": 1,
+   "label": "Basic Rate",
+   "oldfieldname": "basic_rate",
+   "oldfieldtype": "Currency",
+   "options": "Company:company:default_currency",
+   "print_hide": 1
+  },
+  {
+   "default": "0",
+   "fetch_from": "item_code.valuation_rate",
```
Comment (Alchez) [inline erpnext/crm/doctype/opportunity_item/opportunity_item.json]: This shouldn't be the valuation rate, items have a different selling price

**Review by Alchez — CHANGES_REQUESTED** [review]
The logic seems good to me, just do the minor changes on the doctype

**Review by Alchez — APPROVED** [review]

**Pull request merged**